fix: restore the thread's interrupted status after catching InterruptedException (#1005)

micheldavid opened the pull request:
Fixes #1005

elharo commented on google-http-client/src/main/java/com/google/api/client/http/HttpBackOffIOExceptionHandler.java:
```diff
@@ -97,6 +97,8 @@ public boolean handleIOException(HttpRequest request, boolean supportsRetry) thr
     try {
       return BackOffUtils.next(sleeper, backOff);
     } catch (InterruptedException exception) {
+      // Catching InterruptedException clears the thread interrupted bit.
```
I could well be wrong here. I usually am when it comes to Thread arcana, but reading that article and the docs, this only seems necessary when you call `interrupted()`, not when you simply catch InterruptedException. That is, this code does not clear the interrupted bit. Am I missing something?

chanseokoh replied:
This may well be the right thing to do.

Unless this code has control, authority, and context or institutes a policy so that eventually it can make the current interrupted thread die (or decide not to terminate the thread as its normal operation)–which I don't think is the case as a library, or unless this code re-throws `InterruptedException`, or unless it is really correct that this code is designed to consume thread interruption and exit by returning false, I think it should restore the interrupted bit, so that other parts of the code (someone higher in the call stack) can be made aware that this thread is interrupted and act accordingly.

https://stackoverflow.com/questions/2523721/why-do-interruptedexceptions-clear-a-threads-interrupted-status
https://stackoverflow.com/questions/4906799/why-invoke-thread-currentthread-interrupt-in-a-catch-interruptexception-block

But I think the comment `// Catching ... clears the thread interrupted bit.` is unnecessary.

chanseokoh replied:
And I actually think the comment "Catching InterruptedException clears the thread interrupted bit" is incorrect and misleading. It's not the act of catching it. Rather, I believe it is just that the interrupted bit is already cleared (and should by convention) when `InterruptedException` is thrown. From https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html,

> By convention, any method that exits by throwing an InterruptedException clears interrupt status when it does so.

elharo replied:
It would help if this had a test demonstrating both that the bug is real and that this patch fixes it.

micheldavid replied:
Thanks for the quick response and thoughtful discussion.

Sorry about the delay. I havent had spare cycles to work on this recently...
I will write a test, but note that it is difficult to predict threading behavior so the test may be flaky at times, depending on cpu speed and available resources.

elharo replied:
Testing thread issues is extremely tricky. A flaky test can be worse than no test. This might need an integration test that doesn't run as part of the normal suite or some other magic.

elharo commented:
This might help with testing:

https://github.com/awaitility/awaitility

micheldavid commented:
Rebased and added tests. I also re-interrupted the thread in `HttpBackOffUnsuccessfulResponseHandler`.

I noticed [HttpRequest is swallowing an InterruptedException too](https://github.com/googleapis/google-http-java-client/blob/master/google-http-client/src/main/java/com/google/api/client/http/HttpRequest.java#L1071), but that code depends on [BackOffPolicy which javadoc says](https://github.com/googleapis/google-http-java-client/blob/master/google-http-client/src/main/java/com/google/api/client/http/BackOffPolicy.java#L26) is `scheduled to be removed in 1.18` so I did not bother fixing it.

A bit more literature around interrupted exceptions: go/interrupted-exception (internal) and http://www.ibm.com/developerworks/java/library/j-jtp05236/index.html

elharo commented on google-http-client/src/main/java/com/google/api/client/http/HttpBackOffIOExceptionHandler.java:
```diff
@@ -97,6 +97,12 @@ public boolean handleIOException(HttpRequest request, boolean supportsRetry) thr
     try {
       return BackOffUtils.next(sleeper, backOff);
     } catch (InterruptedException exception) {
+<<<<<<< HEAD
```
bad merge

micheldavid replied:
oops... cant believe this got in here without me noticing.
fixed. ptal.

micheldavid commented:
Thanks for the quick turn around. I see a lot of files I did not change are failing the code format lint pass. I did change some of them though. Should I fix it with "mvn com.coveo:fmt-maven-plugin:format"? Maybe fix only the ones I changed?

I also do not have write access, so I will need someone to merge this in.

elharo commented:
yes, please run the plugin

micheldavid commented:
Now the linter is failing on the `google-http-client-bom` project, which doesnt seem to have any code...

From https://source.cloud.google.com/results/invocations/7a0f7d52-8162-4c7b-b711-f7c90e073a42/targets/cloud-devrel%2Fclient-libraries%2Fjava%2Fgoogle-http-java-client%2Fpresubmit%2Flint/log

Failed to execute goal com.coveo:fmt-maven-plugin:2.10:check (default-cli) on project google-http-client-bom

Is this normal? The same seems to happen in master.
I feel like these formatting issues should be resolved in a separate pull request.

chingor13 commented:
Strange, I don't recall us running the code formatter on this repo. Looking into it

chingor13 commented:
We can fix the linting issue separately.

micheldavid commented:
Thanks Jeff!
